File: 03/Perl/t3_2.pl
#!/usr/bin/perl 
#===============================================================================
#
#         FILE: t3_2.pl
#
#        USAGE: ./t3_2.pl  
#
#  DESCRIPTION: 
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: YOUR NAME (), 
# ORGANIZATION: 
#      VERSION: 1.0
#      CREATED: 12/03/2019 09:48:28 AM
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use Data::Dumper;
use 5.020;

my %grid;
my $posx=0;
my $posy=0;
my %steps;
my $steps_count;
my $i;
my $min_count;
my %seen;

sub process_step {

                $grid{$posx}{$posy}++ unless $seen{$posx}{$posy};
                $seen{$posx}{$posy}=1;
                $steps_count++;
                unless ($steps{$i}{$posx}{$posy}) {
                    $steps{$i}{$posx}{$posy} = $steps_count;
                }

                if ($grid{$posx}{$posy} > 1) {
                    my $distance = abs($posx) + abs($posy);
                    my $total_steps = $steps{1}{$posx}{$posy} + $steps{2}{$posx}{$posy};
                    say "intersection $posx $posy distance $distance steps $total_steps";
                    $min_count = $total_steps unless $min_count;
                    $min_count = $total_steps if $min_count > $total_steps;
                    say "min count $min_count";
                }
}

open my $file, '<', '../input' or die 'cannot open input';


$i=0;
while (<$file>) {
    $i++;
    $posx=0;
    $posy=0;
    $steps_count = 0;
    %seen = ();

    chomp;

    my @arr = split /,/, $_;

    for (@arr) {
        my ($direction, $count) = $_ =~ /(.)(.*)/msx;

        if ($direction eq 'R') {
            for (1..$count) {
                $posx++;
                &process_step;
            }
        }

        if ($direction eq 'L') {
            for (1..$count) {
                $posx--;
                &process_step;
            }
        }

        if ($direction eq 'U') {
            for (1..$count) {
                $posy++;
                &process_step;
            }
        }

        if ($direction eq 'D') {
            for (1..$count) {
                $posy--;
                &process_step;
            }
        }

    }

}

File: 03/Python/t3_2.py
#! python3_

from pprint import pprint
import re

grid={}
steps={}

steps_count=0
i = 0
min_count = 0

def process_step(posx,posy) :
                global steps_count
                global steps
                global min_count

                steps_count += 1

                if (i,posx,posy) not in steps:
                    steps[(i,posx,posy)] = steps_count
#                    print("posx "+str(posx)+" posy "+str(posy)+"grid value "+str(grid[(posx,posy)])+" i "+str(i))

                if (grid[(posx,posy)] > 1) :
                    total_steps = steps[(1,posx,posy)] + steps[(2,posx,posy)]
                    distance = abs(posx) + abs(posy)
                    print("intersection "+str(posx)+" "+str(posy)+" distance "+str(distance)+" steps "+str(total_steps))

                    if (min_count == 0) :
                        min_count = total_steps

                    if (min_count > total_steps) :
                        min_count = total_steps

                    print("min count "+str(min_count))

file = open("../input","r")

for line in file:
        i += 1
        arr = line.strip().split(',')
        posx=0
        posy=0
        steps_count = 0
        seen={}
       
        for item in arr:
            match=re.match(r"(.)(.*)", item)
            direction=match.group(1)
            count=match.group(2)
   
            if (direction == 'R') :
                for j in range(1,int(count)+1):
                    posx += 1

                    if (posx,posy) not in seen:
                        if (posx,posy) not in grid:
                            grid[(posx,posy)] = 1
                        else:
                            grid[(posx,posy)] += 1
                    seen[(posx,posy)]=1
                    process_step(posx,posy)
            
            if (direction == 'L') :
                for j in range(1,int(count)+1):
                    posx -= 1

                    if (posx,posy) not in seen:
                        if (posx,posy) not in grid:
                            grid[(posx,posy)] = 1
                        else:
                            grid[(posx,posy)] += 1
                    seen[(posx,posy)]=1
                    process_step(posx,posy)
 
            if (direction == 'U') :
                for j in range(1,int(count)+1):
                    posy += 1

                    if (posx,posy) not in seen:
                        if (posx,posy) not in grid:
                            grid[(posx,posy)] = 1
                        else:
                            grid[(posx,posy)] += 1
                    seen[(posx,posy)]=1
                    process_step(posx,posy)


            if (direction == 'D') :
                for j in range(1,int(count)+1):
                    posy -= 1

                    if (posx,posy) not in seen:
                        if (posx,posy) not in grid:
                            grid[(posx,posy)] = 1
                        else:
                            grid[(posx,posy)] += 1
                    seen[(posx,posy)]=1
                    process_step(posx,posy)


